From a2feba70f39e1c4baa52eae5d68e556255da1a9a Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 27 Aug 2015 00:43:02 -0400
Subject: [PATCH] Optimize hot spot with isValidLocation

While this method is simple, its so hot its showing up in samplings.

Replace with a version much more likely to be inlined that only checks
Y value will speed things up, with no risk when worlds are expected to
never be at the 3 billion distance.

Also static the existing isValidLocation to assist with inlining.
---
 src/main/java/net/minecraft/server/World.java | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 5cbb0dc..999fba9 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -281,7 +281,13 @@ public abstract class World implements IBlockAccess {
         return this.getType(blockposition1).getBlock();
     }
 
-    private boolean isValidLocation(BlockPosition blockposition) {
+    // EMC start
+    private static boolean isFastValidLocation(BlockPosition blockposition) {
+        final int y = blockposition.getY();
+        return y >= 0 && y < 256;
+    }
+    // EMC end
+    private static boolean isValidLocation(BlockPosition blockposition) { // EMC
         return blockposition.getX() >= -30000000 && blockposition.getZ() >= -30000000 && blockposition.getX() < 30000000 && blockposition.getZ() < 30000000 && blockposition.getY() >= 0 && blockposition.getY() < 256;
     }
 
@@ -376,7 +382,7 @@ public abstract class World implements IBlockAccess {
             return true;
         }
         // CraftBukkit end
-        if (!this.isValidLocation(blockposition)) {
+        if (!isFastValidLocation(blockposition)) { // EMC - use a simpler check
             return false;
         } else if (!this.isClientSide && this.worldData.getType() == WorldType.DEBUG_ALL_BLOCK_STATES) {
             return false;
@@ -778,7 +784,7 @@ public abstract class World implements IBlockAccess {
             }
         }
         // CraftBukkit end
-        if (!this.isValidLocation(blockposition)) {
+        if (!this.isFastValidLocation(blockposition)) { // EMC
             return Blocks.AIR.getBlockData();
         } else {
             Chunk chunk = this.getChunkAtWorldCoords(blockposition);
-- 
2.6.2

